src/saveImpl.ts

```diff
@@ -0,0 +1,65 @@
+import * as cache from "@actions/cache";
+import * as core from "@actions/core";
+
+import { Events, Inputs, State } from "./constants";
+import { IStateProvider } from "./stateProvider";
+import * as utils from "./utils/actionUtils";
+
+// Catch and log any unhandled exceptions.  These exceptions can leak out of the uploadChunk method in
+// @actions/toolkit when a failed upload closes the file descriptor causing any in-process reads to
+// throw an uncaught exception.  Instead of failing this action, just warn.
+process.on("uncaughtException", e => utils.logWarning(e.message));
+
+async function saveImpl(stateProvider: IStateProvider): Promise<void> {
+    try {
+        if (!utils.isCacheFeatureAvailable()) {
+            return;
+        }
+
+        if (!utils.isValidEvent()) {
+            utils.logWarning(
+                `Event Validation Error: The event type ${
+                    process.env[Events.Key]
+                } is not supported because it's not tied to a branch or tag ref.`
+            );
+            return;
+        }
+
+        // If restore has stored a primary key in state, reuse that
+        // Else re-evaluate from inputs
+        const primaryKey =
+            stateProvider.getState(State.CachePrimaryKey) ||
+            core.getInput(Inputs.Key);
+
+        if (!primaryKey) {
+            utils.logWarning(`Error retrieving key from state.`);
+            return;
+        }
+
+        // If matched restore key is same as primary key, then do not save cache
+        // NO-OP in case of SaveOnly action
+        const state = stateProvider.getCacheState();
```

Comment:
If my restore was successful in my restore-only action, I would still get `undefined` here right?

Reply:
```
const state = stateProvider.getCacheState() || core.getInput(Inputs.MatchedKey);
```

`MatchKey` will be output from `restore-only` action

Reply:
Just noting things down...

In Save-only action, user can give the following inputs-
1. cache key - from output of `restore` action or same cache `key` as input to `restore` to allow re-evaluation. (can 
be user input or fetched from output of restore-action depending on use case.)
2. cache matched key - this key is the one that was actually restored, could be exact key or restored from restore keys to determine the `cache-hit`. (This has to be taken as input from the output of the previous restore action)
3. path - same as input to restore. (Always has to be inputted by the user)

Reply:
> cache key - from output of restore action or same cache key as input to restore to allow re-evaluation. (can be user input or fetched from output of restore-action depending on use case

@kotewar can we directly read output of earlier step1 with code in step2? Or user still needs to set the input for step2 as output of step1?

Reply:
@bishal-pdMSFT We can if we store it as an environment variable.